knative-operator/pkg/common/service_monitor.go

```diff
@@ -54,23 +81,210 @@ func SetupServiceMonitor(ctx context.Context, cfg *rest.Config, metricsPort int3
-	err = kubemetrics.GenerateAndServeCRMetrics(cfg, ns, filteredGVK, metricsHost, operatorMetricsPort)
+	return nil
+}
+
+func SetupSourceServiceMonitor(client client.Client, namespace string, instance *appsv1.Deployment) error {
+	var sLabel, sNameLabel, sRoleLabel string
+
+	if err := isServiceMonitorAvailable(); err != nil {
+		return err
+	}
+
+	monitor := &monitoringv1.ServiceMonitor{}
+	var SchemeGroupVersion = schema.GroupVersion{Group: "monitoring.coreos.com", Version: "v1"}
+	scheme.Scheme.AddKnownTypes(SchemeGroupVersion, monitor)
+	labels := instance.Spec.Selector.MatchLabels
+
+	if l, ok := labels[SourceLabel]; ok {
+		sLabel = l
+	}
+
+	if l, ok := labels[SourceNameLabel]; ok {
+		sNameLabel = l
+	}
+
+	if l, ok := labels[SourceRoleLabel]; ok {
+		sRoleLabel = l
+	}
+
+	clientOptions := mf.UseClient(mfclient.NewClient(client))
+	// create service for the deployment
+	manifest, err := mf.NewManifest(getMonitorPath(TestSourceServicePath, EventingSourcePath), clientOptions)
 	if err != nil {
+		return fmt.Errorf("unable to parse source service manifest: %w", err)
+	}
+	transforms := []mf.Transformer{updateService(sLabel, sNameLabel, sRoleLabel, instance.Name), mf.InjectOwner(instance), mf.InjectNamespace(namespace)}
+	if manifest, err = manifest.Transform(transforms...); err != nil {
+		return fmt.Errorf("unable to transform source service manifest: %w", err)
+	}
+	if err := manifest.Apply(); err != nil {
+		return err
+	}
+
+	// get service back, needed for the UID and setting owner refs
+	srv := &v1.Service{}
+	if err := client.Get(context.TODO(), types.NamespacedName{Name: instance.Name, Namespace: instance.Namespace}, srv); err != nil {
+		return err
+	}
+	// create service monitor for source
+	manifest, err = mf.NewManifest(getMonitorPath(TestSourceServiceMonitorPath, EventingSourceServiceMonitorPath), clientOptions)
+	if err != nil {
+		return fmt.Errorf("unable to parse source service monitor manifest: %w", err)
+	}
+	transforms = []mf.Transformer{updateServiceMonitor(sLabel, sNameLabel, sRoleLabel, instance.Name), mf.InjectOwner(srv), mf.InjectNamespace(namespace)}
+	if manifest, err = manifest.Transform(transforms...); err != nil {
+		return fmt.Errorf("unable to transform source service monitor manifest: %w", err)
+	}
+	if err := manifest.Apply(); err != nil {
 		return err
 	}
 	return nil
 }
+
+func getMonitorPath(envVar string, defaultVal string) string {
+	path := os.Getenv(envVar)
+	if path == "" {
+		return defaultVal
+	}
+	return path
+}
+
+func updateService(source string, sourceName string, role string, depName string) mf.Transformer {
+	return func(resource *unstructured.Unstructured) error {
+		if resource.GetKind() != "Service" {
+			return nil
+		}
+		var svc = &v1.Service{}
+		if err := scheme.Scheme.Convert(resource, svc, nil); err != nil {
+			return err
+		}
+
+		svc.Name = depName
+		svc.Labels = map[string]string{}
+		svc.Spec.Selector = map[string]string{}
```

Comment:
Are we unsetting stuff here by design? I don't quite understand the code path here but it feels like we should be able to just pass the selector from the deployment as a `map[string]string` and then override here instead of reading the fields and then resetting them here :thinking:

Reply:
Probably I can refactor it. I will check. The idea is to get the service from the manifest and apply the right selector labels.
So Probably I can just pass the deployment selector to both the selector matchlabels and the labs itself. The latter could be avoided but I thought I should add the labels as it is being done in the monitoring services created by the operator-sdk.